### PR TITLE
placement: supports follower role with evict-leader hint

### PR DESCRIPTION
**Pull request by nolouch**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/46304

Problem Summary:

Currently, all `FOLLOWER_CONSTRAINTS` will set the role in the rule to be `voter`. It can cover major scenarios and reduce the user's burden. but in some scenarios cannot cover, like:

there are three az, zone1, zone2, zone3. the user wants the replicas in zone 1 and zone 2 can be the leader, and normal situations, the leader should be placed in zone 1, if zone 1 is gone, the leader is better in zone 2. and zone3 reject be the leader,  zone 3 can be seen as a backup zone, it's not used to serve.

maybe we can enhance the `FOLLOWER_CONSTRAINTS`, like
#####  add attribute
```
Dictionary:  e.g. {"+region=us-east": 2, "+region=us-west,#evict-leader": 1}
```
`#evict-leader` explains voters should be followers, and reject becoming leaders.

full SQL will like this:
```
CREATE PLACEMENT POLICY `mydeploy`  LEADER_CONSTRAINTS='{"+region=us-east-1": 1}' FOLLOWER_CONSTRAINTS='{"+region=us-east-1": 1,  "+region=use-east-2":2, "+region=us-west,#evict-leader": 1};
```

#### alternative plan

**warn**: this is the current easy way to achieve this issue scenario. we would be supported `leader_preferences` in the future to define the leader choice priority.

### What is changed and how it works?

add new keywords.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/placement/rule.go [hunk 1]
```diff
@@ -200,12 +206,29 @@ func NewRules(role PeerRoleType, replicas uint64, cnstr string) ([]*Rule, error)
...
+							overrideRole = Follower
+						}
+					default:
+						return rules, fmt.Errorf("%w: unsupported attribute '%s'", ErrUnsupportedConstraint, str)
+					}
+					continue
+				}
+				newLabels = append(newLabels, str)
+			}
+			labelConstraints, err := NewConstraints(newLabels)
```
Comment (xhebox) [inline ddl/placement/rule.go:226]: LGTM, but what about adding this into `NewConstraints`?

Comment (nolouch) [reply ddl/placement/rule.go:226]: currently, the constraint has no `role` semantics, so I still put it outer the `NewConstraints`.

**Review by xhebox — APPROVED** [review]

**Comment by nolouch** [comment]
PTAL @CabinfeverB

**Review by rleungx — APPROVED** [review]
LGTM, but I was wondering if the user uses # as part of the label, it may have a problem. It's better to add the notice in the doc.

**Pull request merged**